### PR TITLE
feature/5667-theo-DeepLinking

### PR DESCRIPTION
**Pull request by theodur**
## Description of Change

Adds deep linking for secure messaging push notifications. This implementation handles all scenarios of notifications
- When the app is closed
- When the app is opened but in the background
- When the app is opened and in the foreground

The React Navigation linking config was updated to reflect the correct structure of the nav stack. All links with the format `vamobile://messages/:messageID` will navigate the app to the `ViewMessage` screen. The links coming from notifications are expected to be in the `url` field of the `payload`, otherwise the link will not be opened when the notification is tapped. 

I had the VA Notify team update the notification template for secure messaging to add a `url` field to the payload with the path `vamobile://messages/` prepended to the message ID. The field in the payload was previously `msgID`, which would have made it cumbersome to parse the notification and construct the link. It would have also made it difficult to add deep linking for other notifications in the app, since each notification template would have its own custom ID field, and extra logic would need to be added to account for each type of notification. With the `url` field in the notification payload, we only need to check if the field exists, and if so, launch the link. Moving forward, the idea is future push notification templates should include the `url` parameter if they want deep linking to be applicable to them.

The `NotificationManager` is mainly responsible for handling links from incoming notifications. The `registerNotificationOpened` event will launch the link If the user is logged in, otherwise it will store it in the Redux store, and when authentication is complete, `AuthedApp` will launch the link if it's set. The `getInitialNotification` event is called when the closed app is opened from a notification being tapped. Since users have to go through authentication whenever the app is opened from a closed state, the `initialUrl` is automatically stored, which will subsequently launch the link after authentication.

Since Detox supports push notification, I went ahead and added some. Detox was helpful in testing during development ad it provides an easy way to test deep links without going through the SM portal.

#### Notification when app is closed

https://github.com/department-of-veterans-affairs/va-mobile-app/assets/9042882/85e8395c-928f-45c0-92d2-324080199759

#### Notification when app is in background

https://github.com/department-of-veterans-affairs/va-mobile-app/assets/9042882/8d4f2db9-c3ac-434f-ae02-f4997913684f

#### Notification when app is in foreground

https://github.com/department-of-veterans-affairs/va-mobile-app/assets/9042882/ce710fde-a12f-46d3-89b7-5844f7c6e3a1

### Android
Local development mocking a push notification

[PushNotificationAndroid.webm](https://github.com/department-of-veterans-affairs/va-mobile-app/assets/9042882/ce0a9c14-fd46-424f-b16f-46bf66aaee5d)

## Testing

- [x] Tested on iOS 
- [x] Tested on Android 

## Reviewer Validations

Verifying deep links will require access to the SM clinician portal or access to the VA notify API. If you want to test these changes with a real user during review, feel free to reach out to me and I can trigger a push notification from the portal. Using the Detox tests will probably be an easier way to verify/test functionality.

- [ ] Verify SM push notifications navigate the app to the `ViewMessage` screen when the app is closed
- [ ] Verify SM push notifications navigate the app to the `ViewMessage` screen when the app is in the foreground state
- [ ] Verify SM push notifications navigate the app to the `ViewMessage` screen when the app is in the background state
- [ ] Verify non SM push notifications and notifications without the `url` field in their payload don't navigate the app to a specific screen

## PR Checklist

  **Reviewer:** Confirm the items below as you review

- [ ] PR is connected to issue(s)
- [ ] Tests are included to cover this change (when possible)
- [ ] No magic strings (All string unions follow the [Union -> Constant](https://github.com/department-of-veterans-affairs/va-mobile-app/blob/develop/VAMobile/src/constants/common.ts) type pattern)
- [ ] No secrets or API keys are checked in
- [ ] All imports are absolute (no relative imports)
- [ ] New functions and Redux work have proper TSDoc annotations

## For QA

[Run a build for this branch](https://github.com/department-of-veterans-affairs/va-mobile-app/actions/workflows/on_demand_build.yml)

### CODE REVIEW

**Review by Sparowhawk — APPROVED** [review]

File: VAMobile/e2e/tests/PushNotifications.e2e.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    "type": "push"
+  },
+  "title": "New Secure Message",
+  "body": "Review your messages in the health care section of the VA app",
+  "payload": {
+    "url": "vamobile://messages/2092809"
+  },
+}
+
+describe('Push Notifications', () => {
```
Comment (alexandec) [inline VAMobile/e2e/tests/PushNotifications.e2e.ts:16]: Seems like a great use of E2E tests, I like it

File: VAMobile/src/App.tsx [hunk 2]
```diff
@@ -283,12 +284,21 @@ export const AppTabs: FC = () => {
...
   const homeScreens = getHomeScreens()
   const benefitsScreens = getBenefitsScreens()
   const healthScreens = getHealthScreens(useTranslation(NAMESPACE.HEALTH).t)
   const paymentsScreens = getPaymentsScreens()
 
+  // When applicable, this will open the deep link from the notification that launched the app once sign in
+  // is complete. Mapping the link to the appropriate screen is handled by the React Navigation linking config.
+  useEffect(() => {
+    if (initialUrl) {
+      Linking.openURL(initialUrl)
```
Comment (alexandec) [inline VAMobile/src/App.tsx:298]: Would it be useful to clear initialUrl state after navigation completes?

File: VAMobile/src/constants/linking.tsx [hunk 3]
```diff
@@ -12,9 +12,40 @@ export const linking: LinkingOptions<any> = {
...
+            screens: {
+              ViewMessageScreen: 'messages/:messageID',
+            },
+          },
+        },
+      },
     },
   },
+  // Sets the navigation state for deeply nested screens to ensure navigating backwards works correctly
+  getStateFromPath(path) {
```
Comment (alexandec) [inline VAMobile/src/constants/linking.tsx:27]: Will the deep linking tests break if there's a nav change that causes a regression in deep linking? That would be helpful to flag to engineers that they broke something

**Review by alexandec — DISMISSED** [review]
Looks good, just added a couple questions

**Comment by narin** [comment]
@theodur This is looking good for the most part, but I did some testing and I think I found a bug.  It seems that once the Review message screen loads, the body of the message is disappearing. Seems to occur whether the notification is tapped with the app in the foreground or in the background. Please check out these videos:


https://github.com/department-of-veterans-affairs/va-mobile-app/assets/786704/990461be-0935-46b0-8ef5-074b4ce61330


https://github.com/department-of-veterans-affairs/va-mobile-app/assets/786704/49741728-671a-4280-b605-de561125bb61

**Review by narin — CHANGES_REQUESTED** [review]
Fix issue with message bodies disappearing

File: VAMobile/src/constants/linking.tsx (hunk 3 above)
Comment (theodur) [reply VAMobile/src/constants/linking.tsx:27]: Good point, I just added a test for this so people will know if the back ever navigation breaks

**Comment by theodur** [comment]
@narin Should be fixed now ✅

https://github.com/department-of-veterans-affairs/va-mobile-app/assets/9042882/2b8a136b-702e-45fe-8a6a-1c5e6e890d2c

**Review by narin — DISMISSED** [review]

**Review by alexandec — APPROVED** [review]

**Review by dumathane — DISMISSED** [review]

**Review by alexandec — DISMISSED** [review]

**Review by alexandec — APPROVED** [review]

**Review by Sparowhawk — APPROVED** [review]

**Pull request merged**